### ISSUE TITLE
thread-safety considerations for `Table`

### ISSUE THREAD

**Issue by gaukas** [tetratelabs/wazero#1857]
Thank you for your great project. 
 
If I read it correctly the current implementation of `Table` struct in `internal/descriptor/table.go` comes with no thread-safety. 

I would like to know if there exists a reasoning behind this design or any previous related discussion thread. I was trying to implement a feature which requires the Table to be modifiable even after the WebAssembly instance is up and running, it seems reasonable for `Table` to be a concurrent data structure in which case. 

Also, if a pull request adding thread-safety may be accepted I would be happy to implement it in a pull request (maybe even as a non-default feature to be enabled by a build tag).

**Comment by gaukas** [comment]
Also I see the `TODO:` notes. 

https://github.com/tetratelabs/wazero/blob/5796897f37852bf2042cd63959ce4ff673c4366e/internal/sys/fs.go#L253-L261

I am assuming it indicates that it is not intentionally kept thread-unsafe.

**Comment by gaukas** [comment]
Okay, so does this mean wazero will not accept any wasip1-oriented changes regarding concurrent access to these data structures? 

That would be a bummer since such features are somehow prerequisites for socket support in WASI preview 1 (without wasi-socket) with volatile file system context. Thanks for answering my questions tho.

**Comment by ncruces** [comment]
Not sure I understand why? What specifically requires thread safety in sockets?

**Comment by gaukas** [comment]
Correct me if I understood something wrong. 

For example binding file descriptors when the WebAssembly module instance is up and running, at this point the FileTable is accessible to the WebAssembly module already, so it is possible to have shared access?

**Comment by gaukas** [comment]
> What specifically requires thread safety in sockets?

So I guess to address directly to your question: full socket support requires being able to dynamically creating sockets upon WebAssembly's request (i.e., when WebAssembly is up and running), which may cause race conditions between the WebAssembly module and the runtime caller.

**Comment by ncruces** [comment]
This would happen in a host function: WASM calls into the Go host, while this happens the WASM is blocked, waiting for the function to return; no race condition there.

**Comment by gaukas** [comment]
> when this happens the WASM is blocked, waiting for the function to return 

Most likely but there's no strong guarantee? I agree you are right about the WebAssembly-calls-Host model. But consider the (super weird) async model described below: 

We use one socket as control channel plus multiple other sockets for data channels. WebAssembly may asynchronously wait on new notifications from the control channel for "new sockets available". 

In case WebAssembly requested multiple sockets to be created, and receiving notification on one of them. When WebAssembly trying to access the Table, it is possible the Go host is still creating other sockets and pushing to the Table.

**Comment by ncruces** [comment]
No, without WASM/WASI threads, that's not going to happen. Even if the guest exposes that kind of asynchrony, hardware threads are not involved.

**Comment by gaukas** [comment]
I'm not talking about hardware threads. 

Consider we use one goroutine to run a blocking WebAssembly function, which internally uses tokio async to poll on multiple sockets. 

Meanwhile we use another goroutine to push new entries to Table. 

Isn't that concurrent access to Table? One from WebAssembly blocking call, the other from Go host.

**Comment by gaukas** [comment]
An extremist minimal example could be described as

In thread 1, we run a blocking call to a WASM exported function, which reads 32-bit integers (file descriptors) from an existing file descriptor and wrap them into TCPConns. 

In thread 2, we keep pushing new entries into the FileTable. 

Apparently these two threads achieves multithreading and concurrent access to Table through the Go host, which has nothing to do with WASI threading proposal.

**Comment by ncruces** [comment]
Why/how would you use another goroutine to push entries to the table, that's what I'm asking…

You're not supposed to access a runtime from multiple goroutines (concurrently). If you need concurrency at that level, you create _two_ runtimes, and they won't share anything: tables, memory, functions. It may seem that they share the WASI (or other) host module(s), but they don't: the module operates on stuff that's created for each runtime.

And yes, the host is _also_ not supposed to touch runtime stuff from other goroutines while WASM is running. Again: why would it?

**Comment by gaukas** [comment]
Thanks, I guess that just answered my question. Seems what I wanted to achieve is not really encouraged/supported by current WebAssembly standards. In short, we are trying to achieve what a multi-threaded server application can do by exploiting async I/O on sockets. 

The WHY is not easy to fully explain in a few words, but you are welcome to read our paper/project! It is currently going through the process of preprinting and I'm happy to update here once published. 

Thanks again for answering my questions. I will try to explore this a bit further and see if there is a workaround. I appreciate your help.

**Comment by gaukas** [comment]
Btw the preprinted paper as promised, if anyone is interested: https://arxiv.org/abs/2312.00163 😄
